src/main/scala/io/iohk/ethereum/blockchain/sync/SyncStateScheduler.scala

```diff
@@ -0,0 +1,450 @@
+          }
+        }
+    }
+  }
+
+  def processResponse(
+      state: SchedulerState,
+      response: SyncResponse
+  ): Either[ResponseProcessingError, SchedulerState] = {
+    for {
+      activeRequest <- isRequestAlreadyKnownOrResolved(state, response)
+      newState <- processActiveResponse(state, activeRequest, response)
+    } yield newState
+
+  }
+  // scalastyle:off method.length
+  private def createPossibleChildRequests(
+      mptNode: MptNode,
+      parentRequest: StateNodeRequest,
+      requestType: NodeRequest
+  ): Either[NotAccountLeafNode.type, Seq[StateNodeRequest]] = mptNode match {
+    case n: LeafNode =>
+      requestType match {
+        case SyncStateScheduler.StateNode =>
+          Account(n.value).toEither.left.map(_ => NotAccountLeafNode).map { account =>
+            // We are scheduling both storage trie and code requests with highest priority to be sure that leaf nodes completed
+            // as fast as possible
+            val evmRequests = if (account.codeHash != emptyCodeHash) {
+              Seq(StateNodeRequest(account.codeHash, None, Code, Seq(parentRequest.nodeHash), maxMptTrieDepth, 0))
+            } else {
+              Seq()
+            }
+
+            val storageRequests = if (account.storageRoot != emptyStateRootHash) {
+              Seq(
+                StateNodeRequest(
+                  account.storageRoot,
+                  None,
+                  StorageNode,
+                  Seq(parentRequest.nodeHash),
+                  maxMptTrieDepth,
+                  0
+                )
+              )
+            } else {
+              Seq()
+            }
+
+            evmRequests ++ storageRequests
+          }
+
+        case SyncStateScheduler.StorageNode =>
+          Right(Seq())
+      }
+
+    case n: BranchNode =>
+      Right(n.children.collect { case HashNode(childHash) =>
+        StateNodeRequest(
+          ByteString.fromArrayUnsafe(childHash),
+          None,
+          requestType,
+          Seq(parentRequest.nodeHash),
+          parentRequest.nodeDepth + 1,
+          0
+        )
+      })
+
+    case n: ExtensionNode =>
+      Right(n.next match {
+        case HashNode(hash) =>
+          Seq(
+            StateNodeRequest(
+              ByteString(hash),
+              None,
+              requestType,
+              Seq(parentRequest.nodeHash),
+              parentRequest.nodeDepth + n.sharedKey.size,
+              0
+            )
+          )
+        case _ => Nil
+      })
+    case _ => Right(Nil)
+  }
+
+  private def isInDatabase(req: StateNodeRequest): Boolean = {
+    req.requestType match {
+      case request: CodeRequest =>
+        blockchain.getEvmCodeByHash(req.nodeHash).isDefined
+      case request: NodeRequest =>
+        blockchain.getMptNodeByHash(req.nodeHash).isDefined
+    }
+  }
+
+  private def isRequestAlreadyKnown(state: SchedulerState, req: StateNodeRequest): Boolean = {
```

Comment:
Isn't this more of isResponseAlreadyKnown? To difference with `isRequestAlreadyKnownOrResolved`